core/src/main/java/org/apache/calcite/rel/rules/UnionToDistinctRule.java

```diff
@@ -43,7 +43,9 @@
    */
   public UnionToDistinctRule(Class<? extends Union> unionClazz,
       RelBuilderFactory relBuilderFactory) {
-    super(operand(unionClazz, any()), relBuilderFactory, null);
+    super(
+        operandJ(unionClazz, null, union -> !union.all, any()),
+        relBuilderFactory, null);
```

Comment:
Hi @vlsi 
Shall we do analogous optimization for rules like `ProjectSetOpTransposeRule`, `AggregateUnionAggregateRule`, `AggregateUnionTransposeRule`. I mean check `setOp.all` or `union.all` by `operandJ`, rather than checking in method of `onMatch`, thus to save the "match" to be added into optimizer queue

Reply:
That is true: it is better to use `operandJ` or `matches` methods to reduce the planning overhead.

I improved the rules that produced lots of false positives in MaterializedTest, and I have not verified other rules.